### PR TITLE
Slideshow - Add support for maps as textures

### PR DESCRIPTION
**Pull request by PabstMirror**
https://www.youtube.com/watch?v=5J3oPjsDVYQ

adds a helper func that will generate a texture that when loaded will init itself to show a custom map
e.g.
```
call [] call ace_slideshow_fnc_mapImage = 
"#(rgb,4096,4096,1)ui(""ace_slideshow_mapDisplay"",""ace_s_4"")"
```
when that texture is shown it then generates the map image

notes:
- textures are a like a screenshot when first loaded and then never updated
I think it's best to have `showMarkers = 0;` otherwise players may see different things on the same map
e.g. someone jips and will see player markers on the map
- (see slide 3) - sat map hasn't finished loading, same problem can sometimes happen with marker icons
It seems like textures only load if used, and this kind of display might not count as being used?
- computer screen has strange texture problems
- texture is "emmisive" I think? e.g. at full night the whiteboard will be fully brightness, so these really should be on "monitors"
- still need to test in mp and make sure there isn't init order problems

map preload issue https://feedback.bistudio.com/T170918

### CODE REVIEW

**Comment by jonpas** [comment]
What is that example you have there? Looks wrong.

**Comment by PabstMirror** [comment]
working in local mp,
there was a rare chance of it not working because init.sqf is scheduled 
and clients might not finish init before the texture is set by server

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 1]
```diff
@@ -5,12 +5,12 @@
...
  * Arguments:
- * 0: Position <ARRAY>
- * 1: Scale (1.0 fits entire map in x-dimension) <SCALAR>
- * 2: Markers [_pos, _text, _icon, _color] <ARRAY>
- * 3: MapType (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <SCALAR><STRING>
- * 4: Code to run on init (passed [_map, _display, _displayID]) <CODE>
- * 5: Resolution <SCALAR>
+ * 0: Position <ARRAY> (default: center of map)
+ * 1: Scale (1.0 fits entire map in x-dimension) <SCALAR> (default: 1.25)
+ * 2: Markers (each _x = [_pos, _text, _icon, _color]) <ARRAY> (default: [])
```
Comment (jonpas) [inline addons/slideshow/functions/fnc_mapImage.sqf:10]: Split this array out into multiple lines.

Comment (PabstMirror) [reply addons/slideshow/functions/fnc_mapImage.sqf]: requires https://github.com/acemod/ACE3/pull/9303 😄

File: addons/slideshow/GUI.hpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    sizeExLabel = 0;
+    sizeExUnits = 0;
+    sizeExInfo = 0;
+    sizeExLevel = 0;
+    sizeEx = 0;
+    scaleMin = 0.005;
+    scaleMax = 10;
+    showCountourInterval = 0;
+    drawShaded=0.15;
+    shadedSea=0.15;
```
Comment (jonpas) [inline addons/slideshow/GUI.hpp:19]:
```suggestion
    drawShaded = 0.15;
    shadedSea = 0.15;
```

File: addons/slideshow/GUI.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    showMarkers = 0;
+};
+class GVAR(mapTopo): GVAR(mapNormal) {
+    drawShaded=0.35;
+    shadedSea=0.35;
+    sizeExLevel=0.02;
```
Comment (jonpas) [inline addons/slideshow/GUI.hpp:25]:
```suggestion
    drawShaded = 0.35;
    shadedSea = 0.35;
    sizeExLevel = 0.02;
```

File: addons/slideshow/GUI.hpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    colorCountlines[] = {0.647059, 0.533333, 0.286275, 0.5};
+    colorMainCountlines[] = {0.858824, 0, 0,0.5};
+    ptsPerSquareObj = 2000; // don't show buildings
+};
+class GVAR(mapSat): GVAR(mapNormal) {
+    // ref https://feedback.bistudio.com/T170918 - may have problems loading sat textures
+    maxSatelliteAlpha = 0.95;       
+    colorForest[] = {0,1,0,0};
+    colorForestBorder[] = {0,1,0,0};
```
Comment (jonpas) [inline addons/slideshow/GUI.hpp:34]:
```suggestion
    colorForest[] = {0, 1, 0, 0};
    colorForestBorder[] = {0, 1, 0, 0};
```

File: addons/slideshow/XEH_PREP.hpp [hunk 5]
```diff
@@ -1,5 +1,6 @@
-
 PREP(addSlideActions);
 PREP(autoTransition);
 PREP(createSlideshow);
+PREP(mapImage);
+PREP(mapImage_init);
```
Comment (jonpas) [inline addons/slideshow/XEH_PREP.hpp:5]:
```suggestion
PREP(mapImageInit);
```

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 6]
```diff
@@ -0,0 +1,64 @@
+#include "script_component.hpp"
+/*
+ * Author: PabstMirror
+ * Returns a procedural texture that will display a custom map
+ * Needs to be run on all machines
```
Comment (jonpas) [inline addons/slideshow/functions/fnc_mapImage.sqf:5]:
```suggestion
 * Returns a procedural texture that will display a custom map.
 * Needs to be run on all machines.
```

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+ * Example:
+ * [] call ace_slideshow_fnc_mapImage
+ * [getPos player, 0.3, [], 1] call ace_slideshow_fnc_mapImage
+ * [[5000, 5000], 0.5, [[getpos player, "obj"]], 2, {},] call ace_slideshow_fnc_mapImage
+ *
+ * Public: Yes
+ */
+
+params [
+    ["_pos", [worldSize/2,worldSize/2], [[]]],
```
Comment (jonpas) [inline addons/slideshow/functions/fnc_mapImage.sqf:31]:
```suggestion
    ["_pos", [worldSize / 2, worldSize / 2], [[]]],
```

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+ * - 0: Position 2d/3d <ARRAY> (default: [0,0,0])
+ * - 1: Text <STRING> (default: "")
+ * - 2: Marker Type OR icon string (default: "mil_dot")
+ * - 3: Color <ARRAY> (default: [1,0,0,1])
+ * 3: MapType (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <SCALAR><STRING> (default: 0)
+ * 4: Code to run on init (passed [_map, _display, _displayID]) <CODE> (default: {})
+ * 5: Resolution <SCALAR> (default: 4096)
+ *
+ * Return Value:
+ * Texture <STRING>
```
Comment (jonpas) [inline addons/slideshow/functions/fnc_mapImage.sqf:20]:
```suggestion
 * Procedural Texture <STRING>
```

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * Arguments:
+ * 0: Position <ARRAY> (default: center of map)
+ * 1: Scale (1.0 fits entire map in x-dimension) <SCALAR> (default: 1.25)
+ * 2: Markers <ARRAY> (default: [])
+ * - 0: Position 2d/3d <ARRAY> (default: [0,0,0])
+ * - 1: Text <STRING> (default: "")
+ * - 2: Marker Type OR icon string (default: "mil_dot")
+ * - 3: Color <ARRAY> (default: [1,0,0,1])
+ * 3: MapType (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <SCALAR><STRING> (default: 0)
```
Comment (jonpas) [inline addons/slideshow/functions/fnc_mapImage.sqf:15]:
```suggestion
 * 3: Map Type (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <NUMBER, STRING> (default: 0)
```

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+ * Returns a procedural texture that will display a custom map
+ * Needs to be run on all machines
+ *
+ * Arguments:
+ * 0: Position <ARRAY> (default: center of map)
+ * 1: Scale (1.0 fits entire map in x-dimension) <SCALAR> (default: 1.25)
+ * 2: Markers <ARRAY> (default: [])
+ * - 0: Position 2d/3d <ARRAY> (default: [0,0,0])
+ * - 1: Text <STRING> (default: "")
+ * - 2: Marker Type OR icon string (default: "mil_dot")
```
Comment (jonpas) [inline addons/slideshow/functions/fnc_mapImage.sqf:13]:
```suggestion
 * - 2: Marker Type or Icon Name <STRING> (default: "mil_dot")
```

File: addons/slideshow/functions/fnc_createSlideshow.sqf [hunk 11]
```diff
@@ -35,6 +35,7 @@ params [
 if (_objects isEqualTo []) exitWith {
     ERROR("Slideshow Objects field must NOT be empty!");
 };
+if (_names isEqualTo []) then { private _i = 0; _names = _images apply { _i = _i + 1; str _i }};
```
Comment (jonpas) [inline addons/slideshow/functions/fnc_createSlideshow.sqf:38]:
```suggestion
```
I would rather require names in the API. This is not descriptive.

File: docs/wiki/framework/slideshow-framework.md [hunk 12]
```diff
@@ -58,3 +58,58 @@ _Note: Set Name argument added in 3.9.1._
...
+ *
+ * Arguments:
+ * 0: Position <ARRAY> (default: center of map)
+ * 1: Scale (1.0 fits entire map in x-dimension) <SCALAR> (default: 1.25)
+ * 2: Markers (each _x = [_pos, _text, _icon, _color]) <ARRAY> (default: [])
+ * 3: MapType (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <SCALAR><STRING> (default: 0)
+ * 4: Code to run on init (passed [_map, _display, _displayID]) <CODE> (default: {})
+ * 5: Resolution <SCALAR> (default: 4096)
+```
+### 2.2.1 Map Slideshow Example
```
Comment (jonpas) [inline docs/wiki/framework/slideshow-framework.md:78]:
```suggestion

### 2.2.1 Map Slideshow Example

```

File: docs/wiki/framework/slideshow-framework.md [hunk 13]
```diff
@@ -58,3 +58,58 @@ _Note: Set Name argument added in 3.9.1._
...
 5  | `"My Slides"` | Main interaction point name, for easier distinguishing of multiple slideshow sets
 6  | `1` | Uses texture selection 1 for objects with multiple options
+
+### 2.2 Create Map Images
+
+`ace_slideshow_fnc_mapImage`
+
+```sqf
+ * Returns a procedural texture that will display a custom map
+ * Needs to be run on all machines
```
Comment (jonpas) [inline docs/wiki/framework/slideshow-framework.md:68]:
```suggestion
 * Returns a procedural texture that will display a custom map.
 * Needs to be run on all machines.
```

File: docs/wiki/framework/slideshow-framework.md [hunk 14]
```diff
@@ -58,3 +58,58 @@ _Note: Set Name argument added in 3.9.1._
...
+ *
+ * Arguments:
+ * 0: Position <ARRAY> (default: center of map)
+ * 1: Scale (1.0 fits entire map in x-dimension) <SCALAR> (default: 1.25)
+ * 2: Markers (each _x = [_pos, _text, _icon, _color]) <ARRAY> (default: [])
+ * 3: MapType (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <SCALAR><STRING> (default: 0)
+ * 4: Code to run on init (passed [_map, _display, _displayID]) <CODE> (default: {})
+ * 5: Resolution <SCALAR> (default: 4096)
```
Comment (jonpas) [inline docs/wiki/framework/slideshow-framework.md:76]:
```suggestion
 * 0: Position <ARRAY> (default: center of map)
 * 1: Scale (1.0 fits entire map in x-dimension) <NUMBER> (default: 1.25)
 * 2: Markers <ARRAY> (default: [])
 * - 0: Position 2D/3D <ARRAY> (default: [0, 0, 0])
 * - 1: Text <STRING> (default: "")
 * - 2: Marker Type or Icon Name <STRING> (default: "mil_dot")
 * - 3: Color <ARRAY> (default: [1, 0, 0, 1])
 * 3: Map Type (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <NUMBER, STRING> (default: 0)
 * 4: Code to run on init (passed [_map, _display, _displayID]) <CODE> (default: {})
 * 5: Resolution <NUMBER> (default: 4096)
```

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+ * - 3: Color <ARRAY> (default: [1,0,0,1])
+ * 3: MapType (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <SCALAR><STRING> (default: 0)
+ * 4: Code to run on init (passed [_map, _display, _displayID]) <CODE> (default: {})
+ * 5: Resolution <SCALAR> (default: 4096)
```
Comment (jonpas) [inline addons/slideshow/functions/fnc_mapImage.sqf:17]:
```suggestion
 * 5: Resolution <NUMBER> (default: 4096)
```

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+/*
+ * Author: PabstMirror
+ * Returns a procedural texture that will display a custom map
+ * Needs to be run on all machines
+ *
+ * Arguments:
+ * 0: Position <ARRAY> (default: center of map)
+ * 1: Scale (1.0 fits entire map in x-dimension) <SCALAR> (default: 1.25)
+ * 2: Markers <ARRAY> (default: [])
+ * - 0: Position 2d/3d <ARRAY> (default: [0,0,0])
```
Comment (jonpas) [inline addons/slideshow/functions/fnc_mapImage.sqf:11]:
```suggestion
 * - 0: Position 2D/3D <ARRAY> (default: [0, 0, 0])
```

File: docs/wiki/framework/slideshow-framework.md [hunk 17]
```diff
@@ -58,3 +58,58 @@ _Note: Set Name argument added in 3.9.1._
...
+```
+### 2.2.1 Map Slideshow Example
+```sqf
+tex1 = [] call ace_slideshow_fnc_mapImage;
+tex2 = [(getPos aWhiteboard), 0.5, [[getpos aWhiteboard, "you", "mil_start"]], 0] call ace_slideshow_fnc_mapImage;
+tex3 = [[4000, 4000], 0.5, [[[5000, 5000], "target", "mil_objective"]], 2] call ace_slideshow_fnc_mapImage;
+[[aWhiteboard], [], [tex1, tex2, tex3]] call ace_slideshow_fnc_createSlideshow;
```
Comment (jonpas) [inline docs/wiki/framework/slideshow-framework.md:83]:
```suggestion
[[aWhiteboard], ["Full", "Sat Start", "Sat Objective"], [tex1, tex2, tex3]] call ace_slideshow_fnc_createSlideshow;  
```

File: docs/wiki/framework/slideshow-framework.md [hunk 18]
```diff
@@ -58,3 +58,58 @@ _Note: Set Name argument added in 3.9.1._
...
+
+### 2.2.2 Map Slideshow Advanced Example
+```sqf
+private _initCode = {
+   params ["_map", "", "_displayID"];
+
+   // Add custom draw event handler (will only be called when display is actually updated)
+   _map ctrlAddEventHandler ["draw", {
+      params ["_map"];
+      _map drawIcon ["\A3\Drones_F\Air_F_Gamma\UAV_02\Data\UI\Map_UAV_02_CA.paa",[0,0,1,1],getPos theUAV,24,24,getDir theUAV];
```
Comment (jonpas) [inline docs/wiki/framework/slideshow-framework.md:94]:
```suggestion
      _map drawIcon ["\A3\Drones_F\Air_F_Gamma\UAV_02\Data\UI\Map_UAV_02_CA.paa", [0, 0, 1, 1], getPos theUAV, 24, 24, getDir theUAV];
```

File: docs/wiki/framework/slideshow-framework.md [hunk 19]
```diff
@@ -58,3 +58,58 @@ _Note: Set Name argument added in 3.9.1._
...
+        displayUpdate _display;
+        
+        private _map = _display displayCtrl 1;
+        _map ctrlMapAnimAdd [0, 0.3, getpos theUAV];
+        ctrlMapAnimCommit _map;
+    }, 0, _displayID ] call CBA_fnc_addPerFrameHandler;
+};
+
+tex4 = [nil, 0.3, [[getpos aComputer, "you", "mil_start"]], 2, _initCode] call ace_slideshow_fnc_mapImage;
+[[aComputer, aComputer2], [], [tex4]] call ace_slideshow_fnc_createSlideshow;
```
Comment (jonpas) [inline docs/wiki/framework/slideshow-framework.md:114]:
```suggestion
[[aComputer, aComputer2], ["Sat Follow UAV"], [tex4]] call ace_slideshow_fnc_createSlideshow;
```

**Review by jonpas — CHANGES_REQUESTED** [review]

File: docs/wiki/framework/slideshow-framework.md (hunk 17 above)
Comment (PabstMirror) [reply docs/wiki/framework/slideshow-framework.md:83]:
```suggestion
[[aWhiteboard], [tex1, tex2, tex3], ["Full", "Sat Start", "Sat Objective"]] call ace_slideshow_fnc_createSlideshow;  
```

File: docs/wiki/framework/slideshow-framework.md (hunk 19 above)
Comment (PabstMirror) [reply docs/wiki/framework/slideshow-framework.md:114]:
```suggestion
[[aComputer, aComputer2], [], [tex4], ["Sat Follow UAV"]] call ace_slideshow_fnc_createSlideshow;
```

File: addons/slideshow/functions/fnc_mapImage_init.sqf [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * Example:
+ * [findDisplay "ace_s_0"] call ace_slideshow_fnc_mapImage_init
+ *
+ * Public: No
+ */
+
+params [["_display", displayNull], ["_displayID", ""]];
+if (_displayID == "") then { _displayID = displayUniqueName _display; };
+if (isNull _display) then { _display = findDisplay _displayID; };
```
[the remark this answers is not in the document]
Comment (PabstMirror) [reply addons/slideshow/functions/fnc_mapImage_init.sqf:23]:
```
params [["_display", displayNull], ["_displayID", ""]]; // only one of these will be valid
// On primary call from the display's onload we will only have the actual display
if (_displayID == "") then { _displayID = displayUniqueName _display; };
// On the delayed call (due to hash missing) we will only have the displayID
if (isNull _display) then { _display = findDisplay _displayID; };
```

I hope this makes sense, this is a weird way of doing stuff, but I needed a way to rerun this function without messing storing ui vars

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * Arguments:
+ * 0: Position <ARRAY> (default: center of map)
+ * 1: Scale (1.0 fits entire map in x-dimension) <NUMBER> (default: 1.25)
+ * 2: Markers <ARRAY> (default: [])
+ * - 0: Position 2D/3D <ARRAY> (default: [0, 0, 0])
+ * - 1: Text <STRING> (default: "")
+ * - 2: Marker Type or Icon Name <STRING> (default: "mil_dot")
+ * - 3: Color <ARRAY> (default: [1, 0, 0, 1])
+ * 3: Map Type (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <NUMBER, STRING> (default: 0)
```
Comment (rautamiekka) [inline addons/slideshow/functions/fnc_mapImage.sqf:15]:
```suggestion
 * 3: Map Type (0: Normal, 1: Topographic, 2: Satellite) or any custom class (even mission config) <NUMBER, STRING> (default: 0)
```

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    ["_resolution", 4096, [0]]
+];
+TRACE_6("",_pos,_scale,count _markers,_mapType,_userCode isEqualTo {},_resolution);
+
+_markers = _markers apply { // convert marker array to draw command
+    _x params [["_xPos", [0,0,0], [[]]], ["_xText", "", [""]], ["_xIcon", "mil_dot", [""]], ["_xColor", [1,0,0,1], [[]]]];
+    private _size = 0;
+    if (isClass (configFile >> "CfgMarkers" >> _xIcon)) then {
+        _size = 0.5 * getNumber (configFile >> "CfgMarkers" >> _xIcon >> "size");
+        _xIcon = getText (configFile >> "CfgMarkers" >> _xIcon >> "icon"); // don't swap order
```
Comment (rautamiekka) [inline addons/slideshow/functions/fnc_mapImage.sqf:45]:
```suggestion
    private _config = configFile >> "CfgMarkers" >> _xIcon;
    if (isClass _config) then {
        _size = 0.5 * getNumber (_config >> "size");
        _xIcon = getText (_config >> "icon"); // don't swap order
```

File: addons/slideshow/functions/fnc_mapImage_init.sqf [hunk 23]
```diff
@@ -0,0 +1,67 @@
+#include "script_component.hpp"
+/*
+ * Author: PabstMirror
+ * Initilizaes the map texture display.
```
Comment (rautamiekka) [inline addons/slideshow/functions/fnc_mapImage_init.sqf:4]:
```suggestion
 * Initializes the map texture display.
```
To be fair, it's a bit of a hard word to get right.

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+ * Arguments:
+ * 0: Position <ARRAY> (default: center of map)
+ * 1: Scale (1.0 fits entire map in x-dimension) <NUMBER> (default: 1.25)
+ * 2: Markers <ARRAY> (default: [])
+ * - 0: Position 2D/3D <ARRAY> (default: [0, 0, 0])
+ * - 1: Text <STRING> (default: "")
+ * - 2: Marker Type or Icon Name <STRING> (default: "mil_dot")
+ * - 3: Color <ARRAY> (default: [1, 0, 0, 1])
+ * 3: Map Type (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <NUMBER, STRING> (default: 0)
+ * 4: Code to run on init (passed [_map, _display, _displayID]) <CODE> (default: {})
```
Comment (rautamiekka) [inline addons/slideshow/functions/fnc_mapImage.sqf:16]:
```suggestion
 * 4: Code to run on init (will be passed [_map, _display, _displayID]) <CODE> (default: {})
```

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+TRACE_6("",_pos,_scale,count _markers,_mapType,_userCode isEqualTo {},_resolution);
+
+_markers = _markers apply { // convert marker array to draw command
+    _x params [["_xPos", [0,0,0], [[]]], ["_xText", "", [""]], ["_xIcon", "mil_dot", [""]], ["_xColor", [1,0,0,1], [[]]]];
+    private _size = 0;
+    if (isClass (configFile >> "CfgMarkers" >> _xIcon)) then {
+        _size = 0.5 * getNumber (configFile >> "CfgMarkers" >> _xIcon >> "size");
+        _xIcon = getText (configFile >> "CfgMarkers" >> _xIcon >> "icon"); // don't swap order
+    } else {
+        if (_xIcon != "") then { _size = 16 }; // could be file or CfgVehicleIcons
```
Comment (rautamiekka) [inline addons/slideshow/functions/fnc_mapImage.sqf:47]:
```suggestion
        if (_xIcon != "") then { _size = 16 }; // could be a file or a CfgVehicleIcons class
```
If I understood correctly.

File: docs/wiki/framework/slideshow-framework.md [hunk 26]
```diff
@@ -58,3 +58,65 @@ _Note: Set Name argument added in 3.9.1._
...
+ * Arguments:
+ * 0: Position <ARRAY> (default: center of map)
+ * 1: Scale (1.0 fits entire map in x-dimension) <NUMBER> (default: 1.25)
+ * 2: Markers <ARRAY> (default: [])
+ * - 0: Position 2D/3D <ARRAY> (default: [0, 0, 0])
+ * - 1: Text <STRING> (default: "")
+ * - 2: Marker Type or Icon Name <STRING> (default: "mil_dot")
+ * - 3: Color <ARRAY> (default: [1, 0, 0, 1])
+ * 3: Map Type (0: Normal, 1: Topographic, 2: Satelite) or any custom class (even mission config) <NUMBER, STRING> (default: 0)
+ * 4: Code to run on init (passed [_map, _display, _displayID]) <CODE> (default: {})
```
Comment (rautamiekka) [inline docs/wiki/framework/slideshow-framework.md:79]:
```suggestion
 * 4: Code to run on init (will be passed [_map, _display, _displayID]) <CODE> (default: {})
```

**Review by rautamiekka — CHANGES_REQUESTED** [review]

**Review by jonpas — APPROVED** [review]

File: addons/slideshow/functions/fnc_mapImage.sqf [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+    } else {
+        if (_xIcon != "") then { _size = 16 }; // could be a file or a CfgVehicleIcons class
+    };
+    [_xIcon, _xColor, _xPos, _size, _size, 0, _xText, 0, 0.08]
+};
+
+if (_mapType isEqualType 0) then {
+    _mapType = [QGVAR(mapNormal), QGVAR(mapTopo), QGVAR(mapSat)] select _mapType;
+};
+if (!isClass (configFile >> _mapType)) then { ERROR_1("bad map type %1",_mapType); };
```
Comment (PabstMirror) [inline addons/slideshow/functions/fnc_mapImage.sqf:56]:
```suggestion
if ((!isClass (configFile >> _mapType)) && {(!isClass (missionConfigFile >> _mapType))}) then {
    ERROR_1("bad map type %1",_mapType);
};
```

**Comment by godgutten** [comment]
How do we use this ingame?

**Comment by PabstMirror** [comment]
https://ace3.acemod.org/wiki/framework/slideshow-framework#22-create-map-images

**Pull request merged**